homeassistant/components/version/sensor.py

```diff
@@ -72,40 +76,42 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     name = config.get(CONF_NAME)
     source = config.get(CONF_SOURCE)
 
-    session = async_get_clientsession(hass)
-
     channel = HaVersionChannel.BETA if beta else HaVersionChannel.STABLE
+    session = async_get_clientsession(hass)
 
-    if source == "pypi":
-        haversion = VersionData(
-            HaVersion(session, source=HaVersionSource.PYPI, channel=channel)
-        )
-    elif source in ["hassio", "supervisor"]:
-        haversion = VersionData(
-            HaVersion(
-                session, source=HaVersionSource.SUPERVISOR, channel=channel, image=image
-            )
-        )
-    elif source in ["docker", "container"]:
-        if image is not None and image != DEFAULT_IMAGE:
-            image = f"{image}-homeassistant"
-        haversion = VersionData(
-            HaVersion(
-                session, source=HaVersionSource.CONTAINER, channel=channel, image=image
-            )
-        )
-    elif source == "haio":
-        haversion = VersionData(HaVersion(session, source=HaVersionSource.HAIO))
-    else:
-        haversion = VersionData(HaVersion(session, source=HaVersionSource.LOCAL))
-
-    if not name:
-        if source == DEFAULT_SOURCE:
+    if source in HA_VERSION_SOURCES:
+        source = HaVersionSource(source)
+    elif source == "hassio":
+        source = HaVersionSource.SUPERVISOR
+    elif source == "docker":
+        source = HaVersionSource.CONTAINER
+
+    if (
+        source in (HaVersionSource.SUPERVISOR, HaVersionSource.CONTAINER)
+        and image is not None
+        and image != DEFAULT_IMAGE
+    ):
+        image = f"{image}-homeassistant"
+
+    if not (name := config.get(CONF_NAME)):
+        if source == HaVersionSource.LOCAL:
             name = DEFAULT_NAME_LOCAL
         else:
             name = DEFAULT_NAME_LATEST
 
-    async_add_entities([VersionSensor(haversion, name)], True)
+    async_add_entities(
+        [
+            VersionSensor(
+                VersionData(
+                    HaVersion(
+                        session=session, source=source, image=image, channel=channel
+                    )
+                ),
+                SensorEntityDescription(key=source, name=name, icon="mdi:package-up"),
```

Comment:
```suggestion
                SensorEntityDescription(key=source, name=name),
```

The icon is static, so I suggest setting it on `VersionSensor` as a class attribute.